src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.X86.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.ComponentModel.DataAnnotations;
+using System.Runtime.CompilerServices;
+using System.Runtime.Intrinsics.X86;
+using System.Runtime.Intrinsics;
+using System.Runtime.InteropServices;
+
+namespace System.IO.Hashing
+{
+    public partial class Crc32
+    {
+        private const int X86BlockSize = 64;
+
+        private const byte CarrylessMultiplyLower = 0x00;
+        private const byte CarrylessMultiplyUpper = 0x11;
+        private const byte CarrylessMultiplyLeftLowerRightUpper = 0x10;
+
+        // Processes the bytes in source in X86BlockSize chunks using x86 intrinsics, followed by processing 16
+        // byte chunks, and then processing remaining bytes individually. Requires support for Sse2 and Pclmulqdq intrinsics.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static uint UpdateX86(uint crc, ReadOnlySpan<byte> source)
+        {
+            if (source.Length < X86BlockSize)
+            {
+                return UpdateSlowPath(crc, source);
+            }
```

Comment:
Let's move this check to the caller, so that we're not paying for the extra method invocation in this case, and then just assert here that the length is sufficient.

Reply:
Agreed, especially since we're not aggressively inlining anymore.